src/feature/hs_common/shared_random_client.c

```diff
@@ -242,8 +238,14 @@ sr_state_get_start_time_of_current_protocol_run(void)
   networkstatus_t *ns = networkstatus_get_live_consensus(approx_time());
   if (ns) {
     beginning_of_curr_round = ns->valid_after;
+  } else if (authdir_mode(get_options()) || ASSUME_AUTHORITY_SCHEDULING) {
+    beginning_of_curr_round = dirauth_sched_get_cur_valid_after_time();
   } else {
-    beginning_of_curr_round = get_start_time_of_current_round();
+    tor_assert_nonfatal_unreached_once();
```

Comment:
Hm. The soft assert here makes me wonder...

I assume you are soft-asserting here because all the HS-relevant code that calls this function has already checked that the consensus is live before getting n here. That's true. However there is a conceivable TOCTOU race-condition that can trigger this code path; if the consensus was live but about to expire during check, but while moving to this function the second has passed, and now it's not live anymore, so we get to this third state and soft-assert.

Is this something that should concern us? The problem is that if this soft-assert ever triggers we might not be sure if it's the edge case above or a different bug.

Reply:
Hm.  Maybe instead of using a live consensus we should use the latest consensus we have whether it's live or not. Would that help here, do you think?

Reply:
I think that could work indeed. And then we can have a final case where we return the default value and do the soft-assert?

Also we seem to be doing this quadraple if check twice in the code. Can we abstract it somehow?

Reply:
I've tried to make the changes for this on 16f6b65f98a1892b2bca5d781632a49f9ec0bd5f .  Doing so made the two functions diverge slightly, so I'm not sure about the abstraction you suggest, but I think this behavior is more sensible.